### ISSUE TITLE
Integer fancy indexing with nulls behavior

### ISSUE THREAD

**Issue by nsmith-** [scikit-hep/awkward#315]
In the spirit of expanding [integer fancy indexing](https://numpy.org/doc/stable/reference/arrays.indexing.html#integer-array-indexing) to masked arrays, let's review how we got here. We'll use
```
import numpy as np
import pyarrow as pa
import awkward1 as ak
```

For basic numpy arrays,
```
a = np.arange(20).reshape(4, 5)
b = np.array([[0,0], [1,1], [2,2], [3,3]])
c = np.array([[0,1], [2,3], [4,4], [0,0]])
```
we have that
```
>>> a[b, c]
 array([[ 0,  1],
       [ 7,  8],
       [14, 14],
       [15, 15]])
```
i.e. the output has the same shape as `b` and `c` always, and we have to index each dimension of `a` individually.
In awkward, for rectangular arrays, this still holds true:
```
a2 = ak.Array(a)
b2 = ak.Array(b)
c2 = ak.Array(c)
```
gives
```
>>> a2[b2, c2]
 <Array [[0, 1], [7, 8], [14, 14], [15, 15]] type='4 * 2 * int64'>
```

However, awkward does depart from numy as soon as a ListArray is involved. For example, even what appears to be the same array as `c`:
```
cj = ak.Array([[0,1], [2,3], [4,4], [0,0]])
```
we can use `cj` by itself to achieve what previously used both `b` and `c`:
```
>>> ak.type(cj)
 4 * var * int64
>>> a2[cj]
 <Array [[0, 1], [7, 8], [14, 14], [15, 15]] type='4 * var * int64'>
```
Notice `var` in the type of `cj`, which causes the indexing operation to be interpreted differently. We've accepted this as a departure from numpy since the ListArray is a completely different beast and under "our control". Yet, the "shape" of the output continues to match `cj`, i.e. it has the same number of elements in each dimension.

In pyarrow, masked arrays inside indexing operations similarly preserve the "shape" of the indexer, where here we mean the length of the output array (as pyarrow only supports 1D arrays).
```
d = pa.array([0, 1, 2, None, 4, 5])
e = pa.array([0, 2, None, 3])
```
we have
```
>>> d.take(e)
 <pyarrow.lib.Int64Array object at 0x120e1b980>
[
  0,
  2,
  null,
  null
]
```
and awkward faithfully reproduces this:
```
d2 = ak.Array([0, 1, 2, None, 4, 5])
e2 = ak.Array([0, 2, None, 3])
```
gives
```
>>> d2.take(e2)
 <Array [0, 2, None, None] type='4 * ?int64'>
```

Now, combining ListArrays and masked arrays, we continue to preserve the shape of the indexer in the output:
```
f = ak.Array([[0, None, 2], None, [3, 4], []])
g = ak.Array([[1, 2, None], None, [], [None]])
```
gives
```
>>> f[g]
 <Array [[None, 2, None], None, [], [None]] type='4 * option[var * ?int64]'>
```

However, there are a few cases where this breaks down. The following two indexing operations raise an error
```
g2 = ak.Array([[], None, None, []])
f[g2]
```
which should return `[[], None, None, []]`, and
```
g3 = ak.Array([[], [], [], []])
f[g3]
```
which should probably return `[[], [], [], []]` to be consistent with the "same shape", or rather, "trivially broadcastable" property of the output with respect to the indexer. However, I can imagine that `[[], None, [], []]` might be an alternative result.

**Comment by jpivarski** [comment]
The part about

> However, awkward does depart from numy as soon as a ListArray is involved.

is a different topic from the bug reported at

> However, there are a few cases where this breaks down.

The first topic (differing from NumPy) is known and might be called a "gotcha" rather than a bug: we want jagged indexing to be recursive, rather than broadcasted, because we want ufunc'ed predicates to work as jagged boolean slices. The simple cases work as people would expect—it takes some careful digging to realize that it's different from what NumPy does. The argument goes, "If you're paying that close attention, you should also notice the difference between `RegularType` and `ListType`."

The latter part about `f[g2]` and `f[g3]` is actually a bug: None values in the place of lists should count as part of the length of the array. Here is a better example, from our Slack discussion on June 22:

```python
>>> a = ak.Array([[0, 1, 2], [3, 4], [5, 6, 7, 8]])
```

The following should not work:

```python
>>> a[[[False, True, True], [True, False], None, [True, False, False, True]]]
<Array [[1, 2], [3], None, [5, 8]] type='4 * option[var * int64]'>
```

and the following should work:

```python
>>> a[[[False, True, True], None, [True, False, False, True]]]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/pivarski/miniconda3/lib/python3.7/site-packages/awkward1/highlevel.py", line 844, in __getitem__
    return awkward1._util.wrap(self._layout[where], self._behavior)
ValueError: cannot fit jagged slice with length 2 into RegularArray of size 3
```

Basically, the constraint _ought_ to be that the number of top-level elements in the predicate should be equal to the number of top-level elements in the array `a`. Instead, the number of non-None top-level elements in the predicate is constrained to be equal to the number of top-level elements in the array `a`. What _ought_ to happen, after removing the None values and moving down to handle the lists inside, is that the missing lists should be replaced with empty lists at the next level down.

-------

Here's what I said the next morning about how to fix this:

> I was thinking about it this morning. What's probably happening is that project is being called on the option-type array (e.g. IndexedOptionArray), in which None values are removed and you get a non-option-type array of non-None values. There's a variant of this used internally in many places that both projects and keeps track of where the None values were, to put them back in after the operation is carried through on the option-type's content. That's probably why the MissingArraySlice(JaggedArraySlice(...)) acts like a JaggedArraySlice(...) of the non-None values and then puts the Nones back in as placeholders.
> 
> That project-and-reintroduce Nones technique is nice because it's independent of what's inside the option-type array, avoiding an explosion of special cases. But it does the wrong thing here.
> 
> This MissingArraySlice(JaggedArraySlice(...)) is already a special case, and it's good to treat it as such. The other case, MissingArraySlice(ArraySlice()) already works (that's what we determined last night; it's the one Arrow can do) and there aren't any other cases beyond these two: the only Awkward Arrays allowed as slices are option-type (MissingArraySlice(...)), list-type (JaggedArraySlice(...)), and regular, NumPy-like arrays (ArraySlice()). Missing-of-missing shouldn't be allowed—at least, it should be caught and fixed further upstream, not as part of this problem—so the only incorrectly implemented case is MissingArraySlice(JaggedArraySlice(...)). (This includes a possible jagged-of-jagged-of-jagged chain. After the first jagged, the normal machinery can take it the rest of the way down the chain.)
> 
> Supposing we have MissingArraySlice(JaggedArraySlice(...)); what we want to do is not project, but to make a new JaggedArraySlice(...) in which the spots occupied by None are replaced with empty arrays, apply the JaggedArraySlice(...), and then cover the empty array output at the None positions with a ByteMaskedArray that hides them. This is not a project and it explicitly makes use of the knowledge that what the MissingArraySlice contains is a JaggedArraySlice (down one level; it doesn't require knowledge of what's inside of that).
> 
> The "replace None with empty list" strategy only works for integer array slices, but boolean array slices are converted into integer array slices early in the process. Actually passing around boolean array slices is a future optimization that will complicate these implementations. For now, the fact that you'll only ever see integer array slices at this part of the code just means fewer complications to think about.
> 
> It is 100% okay if this requires a new kernel (i.e. custom for loops over the option-type index or the jagged offsets). If you write the for loop in place, in the C++, I can help with the boilerplate of moving that into the cpu-kernels library. We're not bound to minimizing the number of distinct kernels.

**Comment by nsmith-** [comment]
I only mention the ListArray behavior to set up and explain why `f[g]` can even select multiple dimensions (a not-obvious feature from a numpy user's point of view) I don't think its a bug.
I'd like your thoughts on `f[g3]` behavior as that example is more subtle.

**Comment by jpivarski** [comment]
The `f[g2]` and `f[g3]` examples have the additional complication that they involve EmptyArrays, which have floating point type in normal usage (for consistency with NumPy's default), but have integer type when used as a slice (I checked yesterday, thinking that might be relevant for this issue). For user sanity, it should be either integer or Boolean type, since these are the only array types allowed in slices, and if assuming it's Boolean doesn't lead to an error (because of mismatched lengths), assuming it's integer leads to the same answer.

The fact that `f[g3]` converts missing values into empty lists probably shouldn't be allowed. A definitive answer requires more thought—attempting to implement it often provides such clarity—but at the moment, I'm leaning toward no. `f[g2]` is fine.

**Comment by nsmith-** [comment]
In the end,
```
>>> g3 = ak.Array([[], [], [], []])
>>> f[g3]
<Array [[], None, [], []] type='4 * option[var * ?int64]'>
```
was the most straightforward implementation.

**Comment by nsmith-** [comment]
Closed in #322
